### PR TITLE
Try Rules Engine on TestPyPi (was: Actionstest)

### PR DESCRIPTION
**Pull request by thatoldplatitude**
Currently publishing to [test.pypi.org](http://test.pypi.org/) Which I believe we could change to not depend on external python repositories. Also values are hardcoded and are pointing to a package which I control, So builds would fail on the codeforboston repo, as I have it set to only accept pushed builds from my fork.

Just a draft PR of some CI & CD tweaks. This PR introduces building&pushing a wheel to github atrifact, which can only be passed between jobs in a workflow. Then pushing the wheel to pypi registry(in this case the test.pypi.org, whos value is hardcoded fyi.)

Then there are changes to the Dockerfile, to pull down the most current rules-engine wheel from pypi (again, using a hardcoded test.pypi.org value)

This PR shouldn't be merged and is just for transparency. There are other ideas on this build process that would eliminate the need for depending on an external python registry.

N.B. I did not implement any wheel version incrementing in this PR. So if we went with this format we'd need to add automated wheel version incrementing, and reflect those in the pyodide-lock.json file as well.

Peripherally related to #116

### CODE REVIEW

**Comment by thadk** [comment]
Note that this pushes to https://test.pypi.org/project/rules-engine/ whereas there are already quite a number of "rules-engine"'s on the real PyPi : https://pypi.org/search/?q=rules+engine

File: heat-stack/other/.dockerignore [hunk 1]
```diff
@@ -1,9 +1,13 @@
 # This file is moved to the root directory before building the image
 
-/node_modules
+**/node_modules
```
Comment (thadk) [inline heat-stack/other/.dockerignore:3]: Thanks, good trick for monorepos.

**Review by thadk — COMMENTED** [review]
The Dockerfile seemed to work without errors on "production", the push to test PyPi on the Github action did give some errors.

<details>
<summary>detailed output from deploying</summary>


```

#13 [stage-5  2/15] WORKDIR /myapp
#13 DONE 0.0s

#14 [rules-eng 1/1] RUN apt install -y python3 python3-venv && python3 -m venv .venv && source .venv/bin/activate && pip download --no-deps -d .venv/tmp -i https://test.pypi.org/simple/ rules-engine
#0 0.367 
#0 0.367 WARNING: apt does not have a stable CLI interface. Use with caution in scripts.
#0 0.367 
#0 0.369 Reading package lists...
#14 0.926 Building dependency tree...
#14 1.053 Reading state information...
#14 1.180 The following additional packages will be installed:
#14 1.180   krb5-locales libexpat1 libgpm2 libgssapi-krb5-2 libk5crypto3 libkeyutils1
#14 1.181   libkrb5-3 libkrb5support0 libncursesw6 libnsl2 libpython3-stdlib
#14 1.181   libpython3.11-minimal libpython3.11-stdlib libtirpc-common libtirpc3
#14 1.181   media-types python3-distutils python3-lib2to3 python3-minimal
#14 1.181   python3-pip-whl python3-setuptools-whl python3.11 python3.11-minimal
#14 1.181   python3.11-venv
#14 1.182 Suggested packages:
#14 1.182   gpm krb5-doc krb5-user python3-doc python3-tk python3.11-doc binutils
#14 1.182   binfmt-support
#14 1.419 The following NEW packages will be installed:
#14 1.419   krb5-locales libexpat1 libgpm2 libgssapi-krb5-2 libk5crypto3 libkeyutils1
#14 1.419   libkrb5-3 libkrb5support0 libncursesw6 libnsl2 libpython3-stdlib
#14 1.419   libpython3.11-minimal libpython3.11-stdlib libtirpc-common libtirpc3
#14 1.420   media-types python3 python3-distutils python3-lib2to3 python3-minimal
#14 1.420   python3-pip-whl python3-setuptools-whl python3-venv python3.11
#14 1.420   python3.11-minimal python3.11-venv
#14 1.437 0 upgraded, 26 newly installed, 0 to remove and 0 not upgraded.
#14 1.437 Need to get 9412 kB of archives.
#14 1.437 After this operation, 29.5 MB of additional disk space will be used.
#14 1.437 Get:1 http://deb.debian.org/debian bookworm/main amd64 libpython3.11-minimal amd64 3.11.2-6 [813 kB]
#14 1.448 Get:2 http://deb.debian.org/debian bookworm/main amd64 libexpat1 amd64 2.5.0-1 [99.3 kB]
#14 1.449 Get:3 http://deb.debian.org/debian bookworm/main amd64 python3.11-minimal amd64 3.11.2-6 [2064 kB]
#14 1.461 Get:4 http://deb.debian.org/debian bookworm/main amd64 python3-minimal amd64 3.11.2-1+b1 [26.3 kB]
#14 1.462 Get:5 http://deb.debian.org/debian bookworm/main amd64 media-types all 10.0.0 [26.1 kB]
#14 1.462 Get:6 http://deb.debian.org/debian bookworm/main amd64 libncursesw6 amd64 6.4-4 [134 kB]
#14 1.465 Get:7 http://deb.debian.org/debian bookworm/main amd64 libkrb5support0 amd64 1.20.1-2+deb12u1 [32.4 kB]
#14 1.465 Get:8 http://deb.debian.org/debian bookworm/main amd64 libk5crypto3 amd64 1.20.1-2+deb12u1 [78.9 kB]
#14 1.466 Get:9 http://deb.debian.org/debian bookworm/main amd64 libkeyutils1 amd64 1.6.3-2 [8808 B]
#14 1.467 Get:10 http://deb.debian.org/debian bookworm/main amd64 libkrb5-3 amd64 1.20.1-2+deb12u1 [332 kB]
#14 1.469 Get:11 http://deb.debian.org/debian bookworm/main amd64 libgssapi-krb5-2 amd64 1.20.1-2+deb12u1 [134 kB]
#14 1.470 Get:12 http://deb.debian.org/debian bookworm/main amd64 libtirpc-common all 1.3.3+ds-1 [14.0 kB]
#14 1.470 Get:13 http://deb.debian.org/debian bookworm/main amd64 libtirpc3 amd64 1.3.3+ds-1 [85.2 kB]
#14 1.471 Get:14 http://deb.debian.org/debian bookworm/main amd64 libnsl2 amd64 1.3.0-2 [39.5 kB]
#14 1.472 Get:15 http://deb.debian.org/debian bookworm/main amd64 libpython3.11-stdlib amd64 3.11.2-6 [1796 kB]
#14 1.481 Get:16 http://deb.debian.org/debian bookworm/main amd64 python3.11 amd64 3.11.2-6 [572 kB]
#14 1.485 Get:17 http://deb.debian.org/debian bookworm/main amd64 libpython3-stdlib amd64 3.11.2-1+b1 [9312 B]
#14 1.485 Get:18 http://deb.debian.org/debian bookworm/main amd64 python3 amd64 3.11.2-1+b1 [26.3 kB]
#14 1.485 Get:19 http://deb.debian.org/debian bookworm/main amd64 krb5-locales all 1.20.1-2+deb12u1 [62.7 kB]
#14 1.486 Get:20 http://deb.debian.org/debian bookworm/main amd64 libgpm2 amd64 1.20.7-10+b1 [14.2 kB]
#14 1.486 Get:21 http://deb.debian.org/debian bookworm/main amd64 python3-lib2to3 all 3.11.2-3 [76.3 kB]
#14 1.487 Get:22 http://deb.debian.org/debian bookworm/main amd64 python3-distutils all 3.11.2-3 [131 kB]
#14 1.488 Get:23 http://deb.debian.org/debian bookworm/main amd64 python3-pip-whl all 23.0.1+dfsg-1 [1717 kB]
#14 1.501 Get:24 http://deb.debian.org/debian bookworm/main amd64 python3-setuptools-whl all 66.1.1-1 [1111 kB]
#14 1.506 Get:25 http://deb.debian.org/debian bookworm/main amd64 python3.11-venv amd64 3.11.2-6 [5892 B]
#14 1.506 Get:26 http://deb.debian.org/debian bookworm/main amd64 python3-venv amd64 3.11.2-1+b1 [1200 B]
#14 1.619 debconf: delaying package configuration, since apt-utils is not installed
#14 1.638 Fetched 9412 kB in 0s (116 MB/s)
#14 1.650 Selecting previously unselected package libpython3.11-minimal:amd64.
#14 1.650 (Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 6652 files and directories currently installed.)
#14 1.659 Preparing to unpack .../libpython3.11-minimal_3.11.2-6_amd64.deb ...
#14 1.660 Unpacking libpython3.11-minimal:amd64 (3.11.2-6) ...
#14 1.773 Selecting previously unselected package libexpat1:amd64.
#14 1.774 Preparing to unpack .../libexpat1_2.5.0-1_amd64.deb ...
#14 1.775 Unpacking libexpat1:amd64 (2.5.0-1) ...
#14 1.797 Selecting previously unselected package python3.11-minimal.
#14 1.798 Preparing to unpack .../python3.11-minimal_3.11.2-6_amd64.deb ...
#14 1.801 Unpacking python3.11-minimal (3.11.2-6) ...
#14 2.034 Setting up libpython3.11-minimal:amd64 (3.11.2-6) ...
#14 2.042 Setting up libexpat1:amd64 (2.5.0-1) ...
#14 2.045 Setting up python3.11-minimal (3.11.2-6) ...
#14 2.747 Selecting previously unselected package python3-minimal.
#14 2.747 (Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 6968 files and directories currently installed.)
#14 2.754 Preparing to unpack .../00-python3-minimal_3.11.2-1+b1_amd64.deb ...
#14 2.755 Unpacking python3-minimal (3.11.2-1+b1) ...
#14 2.769 Selecting previously unselected package media-types.
#14 2.770 Preparing to unpack .../01-media-types_10.0.0_all.deb ...
#14 2.771 Unpacking media-types (10.0.0) ...
#14 2.783 Selecting previously unselected package libncursesw6:amd64.
#14 2.784 Preparing to unpack .../02-libncursesw6_6.4-4_amd64.deb ...
#14 2.785 Unpacking libncursesw6:amd64 (6.4-4) ...
#14 2.810 Selecting previously unselected package libkrb5support0:amd64.
#14 2.811 Preparing to unpack .../03-libkrb5support0_1.20.1-2+deb12u1_amd64.deb ...
#14 2.812 Unpacking libkrb5support0:amd64 (1.20.1-2+deb12u1) ...
#14 2.825 Selecting previously unselected package libk5crypto3:amd64.
#14 2.826 Preparing to unpack .../04-libk5crypto3_1.20.1-2+deb12u1_amd64.deb ...
#14 2.827 Unpacking libk5crypto3:amd64 (1.20.1-2+deb12u1) ...
#14 2.869 Selecting previously unselected package libkeyutils1:amd64.
#14 2.870 Preparing to unpack .../05-libkeyutils1_1.6.3-2_amd64.deb ...
#14 2.871 Unpacking libkeyutils1:amd64 (1.6.3-2) ...
#14 2.888 Selecting previously unselected package libkrb5-3:amd64.
#14 2.889 Preparing to unpack .../06-libkrb5-3_1.20.1-2+deb12u1_amd64.deb ...
#14 2.890 Unpacking libkrb5-3:amd64 (1.20.1-2+deb12u1) ...
#14 2.932 Selecting previously unselected package libgssapi-krb5-2:amd64.
#14 2.933 Preparing to unpack .../07-libgssapi-krb5-2_1.20.1-2+deb12u1_amd64.deb ...
#14 2.934 Unpacking libgssapi-krb5-2:amd64 (1.20.1-2+deb12u1) ...
#14 2.959 Selecting previously unselected package libtirpc-common.
#14 2.960 Preparing to unpack .../08-libtirpc-common_1.3.3+ds-1_all.deb ...
#14 2.961 Unpacking libtirpc-common (1.3.3+ds-1) ...
#14 2.979 Selecting previously unselected package libtirpc3:amd64.
#14 2.980 Preparing to unpack .../09-libtirpc3_1.3.3+ds-1_amd64.deb ...
#14 2.980 Unpacking libtirpc3:amd64 (1.3.3+ds-1) ...
#14 3.005 Selecting previously unselected package libnsl2:amd64.
#14 3.006 Preparing to unpack .../10-libnsl2_1.3.0-2_amd64.deb ...
#14 3.007 Unpacking libnsl2:amd64 (1.3.0-2) ...
#14 3.021 Selecting previously unselected package libpython3.11-stdlib:amd64.
#14 3.023 Preparing to unpack .../11-libpython3.11-stdlib_3.11.2-6_amd64.deb ...
#14 3.023 Unpacking libpython3.11-stdlib:amd64 (3.11.2-6) ...
#14 3.196 Selecting previously unselected package python3.11.
#14 3.196 Preparing to unpack .../12-python3.11_3.11.2-6_amd64.deb ...
#14 3.197 Unpacking python3.11 (3.11.2-6) ...
#14 3.221 Selecting previously unselected package libpython3-stdlib:amd64.
#14 3.221 Preparing to unpack .../13-libpython3-stdlib_3.11.2-1+b1_amd64.deb ...
#14 3.222 Unpacking libpython3-stdlib:amd64 (3.11.2-1+b1) ...
#14 3.237 Setting up python3-minimal (3.11.2-1+b1) ...
#14 3.395 Selecting previously unselected package python3.
#14 3.395 (Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 7443 files and directories currently installed.)
#14 3.400 Preparing to unpack .../0-python3_3.11.2-1+b1_amd64.deb ...
#14 3.403 Unpacking python3 (3.11.2-1+b1) ...
#14 3.414 Selecting previously unselected package krb5-locales.
#14 3.415 Preparing to unpack .../1-krb5-locales_1.20.1-2+deb12u1_all.deb ...
#14 3.415 Unpacking krb5-locales (1.20.1-2+deb12u1) ...
#14 3.433 Selecting previously unselected package libgpm2:amd64.
#14 3.434 Preparing to unpack .../2-libgpm2_1.20.7-10+b1_amd64.deb ...
#14 3.435 Unpacking libgpm2:amd64 (1.20.7-10+b1) ...
#14 3.447 Selecting previously unselected package python3-lib2to3.
#14 3.448 Preparing to unpack .../3-python3-lib2to3_3.11.2-3_all.deb ...
#14 3.449 Unpacking python3-lib2to3 (3.11.2-3) ...
#14 3.468 Selecting previously unselected package python3-distutils.
#14 3.469 Preparing to unpack .../4-python3-distutils_3.11.2-3_all.deb ...
#14 3.470 Unpacking python3-distutils (3.11.2-3) ...
#14 3.490 Selecting previously unselected package python3-pip-whl.
#14 3.491 Preparing to unpack .../5-python3-pip-whl_23.0.1+dfsg-1_all.deb ...
#14 3.491 Unpacking python3-pip-whl (23.0.1+dfsg-1) ...
#14 3.607 Selecting previously unselected package python3-setuptools-whl.
#14 3.608 Preparing to unpack .../6-python3-setuptools-whl_66.1.1-1_all.deb ...
#14 3.609 Unpacking python3-setuptools-whl (66.1.1-1) ...
#14 3.675 Selecting previously unselected package python3.11-venv.
#14 3.676 Preparing to unpack .../7-python3.11-venv_3.11.2-6_amd64.deb ...
#14 3.677 Unpacking python3.11-venv (3.11.2-6) ...
#14 3.689 Selecting previously unselected package python3-venv.
#14 3.690 Preparing to unpack .../8-python3-venv_3.11.2-1+b1_amd64.deb ...
#14 3.691 Unpacking python3-venv (3.11.2-1+b1) ...
#14 3.711 Setting up media-types (10.0.0) ...
#14 3.716 Setting up python3-setuptools-whl (66.1.1-1) ...
#14 3.718 Setting up libkeyutils1:amd64 (1.6.3-2) ...
#14 3.720 Setting up libgpm2:amd64 (1.20.7-10+b1) ...
#14 3.722 Setting up python3-pip-whl (23.0.1+dfsg-1) ...
#14 3.724 Setting up libtirpc-common (1.3.3+ds-1) ...
#14 3.727 Setting up krb5-locales (1.20.1-2+deb12u1) ...
#14 3.729 Setting up libkrb5support0:amd64 (1.20.1-2+deb12u1) ...
#14 3.731 Setting up libncursesw6:amd64 (6.4-4) ...
#14 3.733 Setting up libk5crypto3:amd64 (1.20.1-2+deb12u1) ...
#14 3.735 Setting up libkrb5-3:amd64 (1.20.1-2+deb12u1) ...
#14 3.737 Setting up libgssapi-krb5-2:amd64 (1.20.1-2+deb12u1) ...
#14 3.740 Setting up libtirpc3:amd64 (1.3.3+ds-1) ...
#14 3.742 Setting up libnsl2:amd64 (1.3.0-2) ...
#14 3.744 Setting up libpython3.11-stdlib:amd64 (3.11.2-6) ...
#14 3.746 Setting up libpython3-stdlib:amd64 (3.11.2-1+b1) ...
#14 3.748 Setting up python3.11 (3.11.2-6) ...
#14 4.542 Setting up python3 (3.11.2-1+b1) ...
#14 4.547 running python rtupdate hooks for python3.11...
#14 4.547 running python post-rtupdate hooks for python3.11...
#14 4.634 Setting up python3-lib2to3 (3.11.2-3) ...
#14 4.734 Setting up python3-distutils (3.11.2-3) ...
#14 4.858 Setting up python3.11-venv (3.11.2-6) ...
#14 4.904 Setting up python3-venv (3.11.2-1+b1) ...
#14 4.906 Processing triggers for libc-bin (2.36-9+deb12u4) ...
#14 9.052 Looking in indexes: https://test.pypi.org/simple/
#14 9.286 Collecting rules-engine
#14 9.404   Downloading https://test-files.pythonhosted.org/packages/5b/99/2d5c9023ec337e2e8ad55ecc6af02a688f737ee3fc3740fb8eca03700619/rules_engine-0.0.1-py3-none-any.whl (7.5 kB)
#14 9.409 Saved /.venv/tmp/rules_engine-0.0.1-py3-none-any.whl
#14 9.409 Successfully downloaded rules-engine
#14 DONE 9.6s

#11 [deps 4/4] RUN npm install --include=dev
#11 14.69 npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs
#11 36.52 
#11 36.52 > postinstall
```

</details>

**Comment by thatoldplatitude** [comment]
> The Dockerfile seemed to work without errors on "production", the push to test PyPi on the Github action did give some errors.
> detailed output from deploying

I did just allow permissions for pushes to test.pypi from codeforboston actions to come thru. It wasn't the fix needed(I don't think. I just added the permissions on the pypi side, then re-triggered the pipeline)
I can check it out tomorrow or sunday tho.

N.B. I would still expect the push to test.pypi to fail due to no version changes on the wheel. But expect the permissions error to be resolved

**Comment by thadk** [comment]
I believe we have integrated all the pieces of this PR indirectly into other branches. We've credited you where possible too :) thanks again. Do you think we've got it all and can close this @thatoldplatitude ?

**Pull request closed without merging**